File: src/main/java/com/github/squirrelgrip/meter/database/hibernate/HibernateDatabaseSession.java
package com.github.squirrelgrip.meter.database.hibernate;

import com.github.squirrelgrip.meter.database.DatabaseSession;
import com.github.squirrelgrip.meter.domain.MeterReading;
import org.hibernate.Session;

public class HibernateDatabaseSession implements DatabaseSession {
    private final Session session;

    public HibernateDatabaseSession(Session session) {
        this.session = session;
    }

    @Override
    public void start() {
        session.beginTransaction();
    }

    @Override
    public void commit() {
        session.getTransaction().commit();
    }

    @Override
    public void rollback() {
        session.getTransaction().rollback();
    }

    @Override
    public void close() {
        session.close();
    }

    @Override
    public void save(MeterReading meterReading) {
        session.persist(meterReading);
    }

    @Override
    public Long count() {
        return (Long) session.createQuery("select count(*) from MeterReading").uniqueResult();
    }
}

File: src/main/kotlin/com/github/squirrelgrip/meter/database/hibernate/HibernateDatabaseSession.kt
package com.github.squirrelgrip.meter.database.hibernate

import com.github.squirrelgrip.meter.database.DatabaseSession
import com.github.squirrelgrip.meter.entity.MeterReading
import org.hibernate.Session

class HibernateDatabaseSession(private val session: Session) : DatabaseSession {
    override fun start() {
        session.beginTransaction()
    }

    override fun commit() {
        session.transaction.commit()
    }

    override fun rollback() {
        session.transaction.rollback()
    }

    override fun close() {
        if (session.isOpen && session.transaction.isActive) {
            rollback()
        }
        session.close()
    }

    override fun save(meterReading: MeterReading) {
        session.persist(meterReading)
    }

    override fun count(): Long =
        session.createQuery("select count(*) from MeterReading", Long::class.java).uniqueResult()

    override fun isTransactionActive(): Boolean =
        session.transaction.isActive
}
